### PR TITLE
Session tracking

### PR DESCRIPTION
**Pull request by hingler**
pull session tracking features into master for beta release

### CODE REVIEW

File: app/src/main/java/studyBuddy/Session.java [hunk 1]
```diff
@@ -27,14 +39,40 @@ public Session() {
...
         productiveTime = 0.0;
         totalTime = 0.0;
-        // assume they were productive the whole time
         percentProductive = 1.0;
-        expectedTime = 0.0;
-        // timeManager = null;
         sessionOngoing = false;
+        sessionPaused = false;
+
+        // manage session events
```
Comment (georgina-chandler) [inline app/src/main/java/studyBuddy/Session.java:49]: can we make the comments on these a bit longer - it took me a long time to trace them around

File: app/src/main/java/studyBuddy/Session.java [hunk 2]
```diff
@@ -47,13 +85,70 @@ public Session() {
...
+
+            runner.setCallback(callback);
+            runner.setStartTime(this.startTime.getTime());
+            runner.setDuration(expectedSessionTime);
+            // the runner and the session are now synchronized
+            handler.post(runner);
+        }
+    }
+
+    public void startSession(String sessionName, long expectedSessionTime) {
```
Comment (georgina-chandler) [inline app/src/main/java/studyBuddy/Session.java:112]: comment explaining why we need this?

File: app/src/main/java/studyBuddy/timemanagement/EndSessionButtonListener.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+import android.view.View;
+
+import com.example.studdybuddy.R;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import studyBuddy.Session;
+
+// TODO: change onTouchListener to onClickListener
+public class EndSessionButtonListener implements View.OnTouchListener {
```
Comment (georgina-chandler) [inline app/src/main/java/studyBuddy/timemanagement/EndSessionButtonListener.java:18]: does this need to be done

File: app/src/main/java/studyBuddy/timemanagement/EndSessionButtonListener.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        animator.setTarget(target);
+        runnable = () -> {
+            animator.reverse();
+            isOpen.set(false);
+        };
+    }
+
+    @Override
+    public boolean onTouch(View v, MotionEvent e) {
+        v.performClick();
```
Comment (georgina-chandler) [inline app/src/main/java/studyBuddy/timemanagement/EndSessionButtonListener.java:49]: this gesture seems kind of complicated - fine for beta though

**Review by georgina-chandler — COMMENTED** [review]
Do you know why build is failing?
Looks good overall just would be nice to have more comments for people coming into these files.
Do we need to change from touch to click now?

File: app/src/main/java/studyBuddy/timemanagement/EndSessionButtonListener.java (hunk 3 above)
Comment (hingler) [reply app/src/main/java/studyBuddy/timemanagement/EndSessionButtonListener.java:18]: not atm, there's some additional functionality to handle the difference between the two that we can remove by making the change

File: app/src/main/java/studyBuddy/Session.java (hunk 1 above)
Comment (hingler) [reply app/src/main/java/studyBuddy/Session.java:49]: for sure -- i'll take a pass thru all of this and add some more detailed comments to everything

**Comment by jamieboy1337** [comment]
mockito tests do not run on 1.11 -- our project is currently running on 1.8 (can't go any higher on android) and it works just fine once we've bumped to that version of java

**Comment by haciim** [comment]
Looks good to me

**Review by haciim — APPROVED** [review]

**Comment by hingler** [comment]
> 
> 
> Do you know why build is failing?

Not 100 percent sure why build is failing but it definitely has something to do with a version inconsistency. currently our project is being built on java 1.8 but the build script was running on 1.11, this seemed to work fine for our tests but the timer runner uses mockito to fake some android library functionality and we must have run into an issue there.

> Looks good overall just would be nice to have more comments for people coming into these files.

Before we merge I'll make sure to go through and freshen up the comments. Apologies for getting lazy there, a lot of the classes are severely lacking in documentation

> Do we need to change from touch to click now?

Not atm, it's something I wanted to make a note of since I wasn't aware that it existed

**Pull request merged**